File: runtimes/swiftui/Sources/StyleConverterRuntime/Renderer/VerticalBlockFlowLayout.swift
//
//  VerticalBlockFlowLayout.swift
//  Renderer — wave 47 (lane Z2).
//
//  The vertical-writing-mode BLOCK-FLOW seam (css-writing-modes-4 §6: under
//  `vertical-rl`/`vertical-lr` the block-flow direction is HORIZONTAL, so a
//  block container's in-flow children stack side-by-side — right→left for
//  vertical-rl, left→right for vertical-lr — instead of the VStack that is
//  correct for horizontal-tb). Companion of VerticalTextFlowLayout.swift
//  (the inline-axis half); consumed only by ComponentRenderer's block
//  branch, behind the Z2 capture gate documented there. The Compose twin is
//  core/renderer/VerticalBlockFlowLayout.kt — same placement table
//  (VerticalBlockFlowMathTest pins it).
//

// SwiftUI for the Layout protocol — same availability discipline as
// FloatBlockLayout / InlineAtomBlockLayout — and the environment key that
// marks the scope the vertical fill folds are allowed to act in.
import SwiftUI

/// Wave 47 (lane Z2) — true only INSIDE a flow this lane actually models:
/// the VerticalBlockFlowLayout seam's content and a vertical multicol
/// fragment column (both publish it in ComponentRenderer). The renderer's
/// vertical fill decisions — suppressing the §10.3.3 WIDTH fill and
/// applying the inline-axis (height) stretch — are gated on it, so every
/// vertical box still rendered by the FROZEN horizontal-tb paths (legacy
/// VStacks, flex/grid items — the css-writing-modes available-size family,
/// which PASSES on those paths today) keeps its pre-Z2 fills
/// byte-identically. Compose twin: LocalVerticalFlowScope.
private struct VerticalFlowScopeZ2Key: EnvironmentKey {
    static let defaultValue: Bool = false
}

extension EnvironmentValues {
    /// See VerticalFlowScopeZ2Key's doc above.
    var verticalFlowScopeZ2: Bool {
        get { self[VerticalFlowScopeZ2Key.self] }
        set { self[VerticalFlowScopeZ2Key.self] = newValue }
    }
}

/// Block-flow container for a VERTICAL writing mode: subviews measure at
/// their IDEAL width (a box's width is its BLOCK size — own declaration or
/// content — never a stretch; the §10.3.3 stretch fit belongs to the inline
/// axis, i.e. the height here) and place sequentially along the horizontal
/// axis. §8.3.1 margin COLLAPSE along this axis is NOT modeled — the
/// caller logs that honestly; margins ride inside each subview's own box
/// (the repo's padding-based margin emulation), so plain summing reproduces
/// single-sided block-axis margins exactly.
@available(iOS 16.0, *)
struct VerticalBlockFlowLayout: Layout {
    /// true = vertical-rl / sideways-rl: the FIRST child sits at the
    /// container's RIGHT edge (block-start, §6.4) and flow walks leftward.
    let blockRtl: Bool

    /// Per-subview proposal: ideal width (nil = "whatever you want", the
    /// box's own block size), the container's proposed height as the
    /// inline-axis cap (the stretch-fit basis the vertical fill fold
    /// targets statically).
    private func childProposal(_ proposal: ProposedViewSize) -> ProposedViewSize {
        ProposedViewSize(width: nil, height: proposal.height)
    }

    func sizeThatFits(proposal: ProposedViewSize,
                      subviews: Subviews,
                      cache: inout ()) -> CGSize {
        // The container's used block extent = the SUM of child extents
        // (css-writing-modes-4 §7.3 — an orthogonal box's auto block size
        // hugs its content, even past the ICB); its inline extent = the
        // proposal (the style chain's height pin) or the tallest child.
        let sizes = subviews.map { $0.sizeThatFits(childProposal(proposal)) }
        let total = sizes.reduce(CGFloat(0)) { $0 + $1.width }
        let maxH = sizes.map(\.height).max() ?? 0
        return CGSize(width: total, height: proposal.height ?? maxH)
    }

    func placeSubviews(in bounds: CGRect,
                       proposal: ProposedViewSize,
                       subviews: Subviews,
                       cache: inout ()) {
        // Re-measure with the same proposal the sizing pass used so the
        // two passes can never drift (the CSSFlexLayout discipline).
        let sizes = subviews.map { $0.sizeThatFits(childProposal(proposal)) }
        // The anchor: bounds.maxX IS the container's block-start edge for
        // rl — for an auto-sized box bounds.width is our reported content
        // sum, and for an author-pinned narrower box overflow then escapes
        // past block-end (LEFT), exactly the browser's escape direction
        // (Compose twin: VerticalBlockFlowMath.positions' anchor rule).
        var x = blockRtl ? bounds.maxX : bounds.minX
        for (index, subview) in subviews.enumerated() {
            let w = sizes[index].width
            // rl walks leftward from the right edge; lr rightward from 0.
            if blockRtl { x -= w }
            // In-flow block boxes share the container's inline-start
            // (top) edge; each keeps its own measured size.
            subview.place(
                at: CGPoint(x: x, y: bounds.minY),
                anchor: .topLeading,
                proposal: ProposedViewSize(width: w, height: sizes[index].height))
            if !blockRtl { x += w }
        }
    }
}

File: runtimes/compose/src/main/java/com/styleconverter/runtime/core/renderer/VerticalBlockFlowLayout.kt
// Wave-47 lane Z2 — the vertical-writing-mode BLOCK-FLOW seam
// (css-writing-modes-4 §6: under `vertical-rl`/`vertical-lr` the block-flow
// direction is HORIZONTAL, so a block container's in-flow children stack
// side-by-side — right→left for vertical-rl, left→right for vertical-lr —
// instead of the Column stack that is correct for horizontal-tb).
// Companion of VerticalTextFlowLayout.kt (the inline-axis half of vertical
// writing modes); consumed only by ComponentRenderer's block branch, behind
// the Z2 capture gate documented there.
package com.styleconverter.runtime.core.renderer

// Composable + Layout are the minimal Compose surface: one custom measure
// policy, no state, no drawing; the compositionLocal marks the scope the
// vertical fill folds are allowed to act in.
import androidx.compose.runtime.Composable
import androidx.compose.runtime.compositionLocalOf
import androidx.compose.ui.Modifier
import androidx.compose.ui.layout.Layout
import androidx.compose.ui.unit.Constraints

/**
 * Wave-47 lane Z2 — true only INSIDE a flow this lane actually models:
 * the [VerticalBlockFlowLayout] seam's content and a vertical multicol
 * container's content (both provide it in ComponentRenderer). The
 * renderer's vertical fill decisions — suppressing the §10.3.3 WIDTH fill
 * and applying the inline-axis (height) stretch — are gated on it, so
 * every vertical box still rendered by the FROZEN horizontal-tb paths
 * (legacy Columns, flex/grid items — the css-writing-modes
 * available-size family, which PASSES on those paths today) keeps its
 * pre-Z2 fills byte-identically.
 */
internal val LocalVerticalFlowScope = compositionLocalOf { false }

/**
 * The pure placement half, split from the composable so the JVM suite can
 * pin the position table without a Compose host (same split as the multicol
 * S/V-tables).
 */
internal object VerticalBlockFlowMath {

    /**
     * X origin of each child box, in child order.
     *
     * @param widths      each child's measured width (its margin box — the
     *        margin rides inside the measurable as `absolutePadding`, the
     *        repo's margin emulation, so plain summing reproduces block-axis
     *        margin placement; adjoining-margin COLLAPSE along this axis is
     *        NOT modeled — the caller logs that honestly).
     * @param anchorWidth the container's used block extent: the sum of child
     *        widths for an auto-sized container (css-writing-modes-4 §7.3 —
     *        the block size of the orthogonal box hugs its content, even
     *        past the ICB), or the constrained width when the author pinned
     *        it (overflow then escapes past block-end, i.e. LEFT for
     *        vertical-rl — exactly what anchoring at the fixed right edge
     *        produces).
     * @param blockRtl    true for vertical-rl / sideways-rl: the first child
     *        sits at the container's RIGHT edge (block-start, §6.4) and flow
     *        walks leftward; false (vertical-lr) walks left→right from 0.
     */
    fun positions(widths: List<Int>, anchorWidth: Int, blockRtl: Boolean): List<Int> {
        // Running block-axis cursor: distance consumed before each child.
        var consumed = 0
        return widths.map { w ->
            val x = if (blockRtl) anchorWidth - consumed - w else consumed
            consumed += w
            x
        }
    }
}

/**
 * Block-flow container for a VERTICAL writing mode: children stack along
 * the horizontal axis in [VerticalBlockFlowMath.positions] order. Children
 * measure with a LOOSE, unbounded width (their own block size — width here
 * — is theirs to decide, css-writing-modes-4 §7.3) and the container's
 * inline extent as max height (the stretch-fit basis the renderer's
 * vertical fill fold targets).
 */
@Composable
internal fun VerticalBlockFlowLayout(
    // true = vertical-rl / sideways-rl (block axis right→left).
    blockRtl: Boolean,
    // The container's own style chain (size, background, margins…).
    modifier: Modifier = Modifier,
    content: @Composable () -> Unit
) {
    Layout(content = content, modifier = modifier) { measurables, constraints ->
        // Loose width: a child's width is its BLOCK size (own declaration or
        // content), never a stretch — the §10.3.3 stretch fit belongs to the
        // INLINE axis, which is the height cap we keep from the incoming
        // constraints (min loosened so non-filling children may be shorter).
        val childConstraints = constraints.copy(
            minWidth = 0, maxWidth = Constraints.Infinity, minHeight = 0)
        val placeables = measurables.map { it.measure(childConstraints) }
        // The container's used block extent = the sum of child extents
        // (content-hugging §7.3); its inline extent = the tallest child
        // unless the style chain pinned it (tight constraints win below).
        val totalW = placeables.sumOf { it.width }
        val maxH = placeables.maxOfOrNull { it.height } ?: 0
        // Compose requires the reported size inside the constraints — an
        // over-wide orthogonal box reports the cap and lets children draw
        // past it (block boxes do not clip by default, matching CSS
        // overflow: visible).
        val layoutW = totalW.coerceIn(constraints.minWidth, constraints.maxWidth)
        val layoutH =
            if (constraints.hasFixedHeight) constraints.maxHeight
            else maxH.coerceIn(constraints.minHeight, constraints.maxHeight)
        // Anchor (see VerticalBlockFlowMath.positions): the TRUE content
        // extent for an auto-width container — placement must not collapse
        // to the clamped report or vertical-rl would show the wrong end of
        // an over-wide flow — and the fixed edge when the width is pinned.
        val anchor = if (constraints.hasFixedWidth) layoutW else maxOf(totalW, layoutW)
        val xs = VerticalBlockFlowMath.positions(
            placeables.map { it.width }, anchor, blockRtl)
        layout(layoutW, layoutH) {
            // Every child at its block-axis slot, inline origin 0 (in-flow
            // block boxes share the container's inline start edge).
            placeables.forEachIndexed { i, p -> p.place(xs[i], 0) }
        }
    }
}
